renku/cli/_graph.py

```diff
@@ -67,62 +67,55 @@ def __attrs_post_init__(self):
     @_nodes.default
     def default_nodes(self):
         """Build node index."""
-        _nodes = OrderedDict()
-        _need_update = self._need_update = {}
-        _latest_commits = self._latest_commits
+        nodes = []
+        for commit in reversed(self._sorted_commits):
+            try:
+                activity = self.commits[commit]
+                nodes.extend(reversed(list(activity.nodes)))
+            except KeyError:
+                pass
+        return nodes
 
-        def _update_node(node, need_update, store_node=True):
-            path = node.path
-            latest = _latest_commits.get(path)
+    def need_update(self, node):
+        """Return out-dated nodes."""
+        if node._id in self._need_update:
+            return self._need_update[node._id]
 
-            if latest is None:
+        latest = self.latest(node)
+        if latest:
+            return self._need_update.setdefault(node._id, [node])
+
+        need_update_ = []
+        for parent in node.parents:
+            parent_updates = self.need_update(parent)
+            if parent_updates:
+                need_update_.extend(parent_updates)
+
+        return self._need_update.setdefault(node._id, need_update_)
+
+    def latest(self, node):
+        """Check if the node is in the latest version."""
```

Comment:
The docstring implies that the return value should be a boolean, but instead it's the latest commit for this node.